onnxruntime/core/providers/webnn/builders/impl/gemm_op_builder.cc

```diff
@@ -162,6 +157,29 @@ bool GemmOpBuilder::IsOpSupportedImpl(const InitializedTensorSet& initializers,
     }
   }
 
+  if (op_type == "MatMul") {
+    if (a_shape.size() < 2 || b_shape.size() < 2) {
+      LOGS(logger, VERBOSE) << "Inputs of MatMul must be at least 2D";
+      return false;
+    }
+
+    // WebNN CPU backend has two more constraints.
+    // https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/modules/ml/webnn/ml_graph_xnnpack.cc;l=1177
+    if (device_type == WebnnDeviceType::CPU) {
+      if (a_shape.size() != b_shape.size()) {
+        LOGS(logger, VERBOSE) << "The rank of two inputs for WebNN CPU backend MatMul must be the same.";
+        return false;
+      }
+
+      for (size_t i = 0; i < a_shape.size() - 2; i++) {
+        if (a_shape[i] != b_shape[i]) {
+          LOGS(logger, VERBOSE) << "WebNN CPU backend can't support broadcasting for MatMul.";
```

Comment:
Same here - if the backend can't handle it, then it's not conformant to the spec, and the Chromium backend needs to call `expand` before passing it to XNNPack's MatMul. So we should add a "TODO: " here too. Otherwise this will just mask the bug in Chromium that other people will hit when they call WebNN directly.